### PR TITLE
Reusing copier for AVRO schemas

### LINKED ISSUE #162: Implement Avro Schema Copier for S3-S3 replications
_As a user_
_I'd like to implement a new schema copier_
_So that I can copy external avro schema from s3 to another s3 location_

Our current implementation of the SchemaCopier (https://github.com/HotelsDotCom/circus-train/blob/master/circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java) Assumes the target is HDFS and doesn't work correctly when trying to replicate data from s3 to another s3 location.

### Acceptance Criteria
* Implement an s3s3SchemaCopier that applies the same settings as the data copier (e.g. canned-acl, same configured CredentialProvider.

### PR DESCRIPTION
**Pull request by patduin**
The idea is to reuse the Copier for the avro schema. Treating it as a data file benefit is that any fixes we do in the copier (regarding ACL, credentials or anything) will be picked up automatically. The only downside I can think of is custom copiers not expecting to copy a schema file. But perhaps that's something we can live with or can be adjusted fairly easily in custom third party copiers.

fixes #162

### CODE REVIEW

File: circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolverTest.java [hunk 1]
```diff
@@ -73,7 +81,7 @@ public void resolveNameServicesAddsAuthorityToPathWithoutScheme() {
     setDfsPaths("foo");
     Path input = new Path("/etl/test/avsc/schema.avsc");
     Path result = resolver.resolveNameServices(input);
-    Path expected = new Path("/foo/etl/test/avsc/schema.avsc");
+    Path expected = new Path("/etl/test/avsc/schema.avsc");
```
Comment (patduin) [inline circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolverTest.java]: @abhimanyugupta07  does this seem correct to you?

File: circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/SchemaCopierTest.java [hunk 2]
```diff
@@ -26,21 +26,39 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
 
+import com.hotels.bdp.circustrain.api.copier.CopierFactoryManager;
+import com.hotels.bdp.circustrain.api.copier.CopierOptions;
+import com.hotels.bdp.circustrain.api.event.EventTableReplication;
+
+@RunWith(MockitoJUnitRunner.class)
 public class SchemaCopierTest {
```
Comment (patduin) [inline circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/SchemaCopierTest.java]: This test is broken I'll fix it...

File: circus-train-s3-mapreduce-cp-copier/src/test/java/com/hotels/bdp/circustrain/s3mapreducecpcopier/S3MapReduceCpCopierTest.java [hunk 3]
```diff
@@ -104,23 +106,41 @@ public void tableArgsAndConfiguration() throws Exception {
...
-    when(copierOptions.get(MULTIPART_UPLOAD_THRESHOLD)).thenReturn("123456");
-    when(copierOptions.get(MAX_MAPS)).thenReturn("78");
-    when(copierOptions.get(COPY_STRATEGY)).thenReturn("the-strategy");
-    when(copierOptions.get(LOG_PATH)).thenReturn("hdfs://path/to/logs/");
-    when(copierOptions.get(REGION)).thenReturn("us-east-1");
-    when(copierOptions.get(IGNORE_FAILURES)).thenReturn("true");
-    when(copierOptions.get(CANNED_ACL)).thenReturn(CannedAccessControlList.BucketOwnerFullControl.toString());
-
-    S3MapReduceCpCopier copier = new S3MapReduceCpCopier(conf, sourceDataBaseLocation, Collections.<Path> emptyList(),
+    copierOptions.put(CREDENTIAL_PROVIDER, "jceks://hdfs/foo/bar.jceks");
```
Comment (patduin) [inline circus-train-s3-mapreduce-cp-copier/src/test/java/com/hotels/bdp/circustrain/s3mapreducecpcopier/S3MapReduceCpCopierTest.java]: mocking a Map, madness!

File: circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/transformation/AbstractAvroSerDeTransformation.java [hunk 4]
```diff
@@ -85,8 +91,15 @@ public void tableReplicationStart(EventTableReplication tableReplication, String
...
+    this.eventId = null;
+    this.tableReplication = null;
+  }
 
   @Override
-  public void tableReplicationFailure(EventTableReplication tableReplication, String eventId, Throwable t) {}
+  public void tableReplicationFailure(EventTableReplication tableReplication, String eventId, Throwable t) {
+    this.eventId = null;
+    this.tableReplication = null;
+
```
Comment (massdosage) [inline circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/transformation/AbstractAvroSerDeTransformation.java]: remove blank line

File: circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolver.java [hunk 5]
```diff
@@ -58,21 +58,22 @@ public Path resolveScheme(Path path) {
...
-      final String original = path.toString();
-      if (isBlank(scheme)) {
-        url = String.format("/%s%s", nameService, path);
-        path = new Path(url);
-      } else {
-        path = new Path(scheme, nameService, url);
+    if ("hdfs".equalsIgnoreCase(path.toUri().getScheme())) {
+      String nameService = configuration.get(DFSConfigKeys.DFS_NAMESERVICES);
+      if (isNotBlank(nameService)) {
+        URI uri = path.toUri();
```
Comment (massdosage) [inline circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolver.java]: NABD but you could consider moving these out of the if statement and then re-using them in both the `if()` clause and here.

File: circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java [hunk 6]
```diff
@@ -41,64 +44,42 @@
...
-    try {
-      destinationFileSystem = remoteDestinationLocation.getFileSystem(replicaHiveConf);
-      destinationFileSystem.copyFromLocalFile(localLocation, remoteDestinationLocation);
-    } catch (IOException e) {
-      throw new CircusTrainException("Couldn't copy file from " + localLocation + " to " + remoteDestinationLocation,
-          e);
-    }
+    Path destinationSchemaFile = new Path(destination, sourceLocation.getName());
+
+    Map<String, Object> mergeCopierOptions = new HashMap<>(TableReplication
```
Comment (massdosage) [inline circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java]:
```suggestion
    Map<String, Object> mergedCopierOptions = new HashMap<>(TableReplication
```

File: circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java [hunk 7]
```diff
@@ -41,64 +44,42 @@
...
+        .getMergedCopierOptions(globalCopierOptions.getCopierOptions(), eventTableReplication.getCopierOptions()));
+    mergeCopierOptions.put(CopierOptions.COPY_DESTINATION_IS_FILE, "true");
+    CopierFactory copierFactory = copierFactoryManager
+        .getCopierFactory(sourceLocation, destinationSchemaFile, mergeCopierOptions);
+    LOG.info("Going to replicate the Avro schema from '{}' to '{}'", sourceLocation, destinationSchemaFile);
```
Comment (massdosage) [inline circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java]:
```suggestion
    LOG.info("Replicating Avro schema from '{}' to '{}'", sourceLocation, destinationSchemaFile);
```

File: circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java [hunk 8]
```diff
@@ -41,64 +44,42 @@
...
+    Copier copier = copierFactory.newInstance(eventId, sourceLocation, destinationSchemaFile, mergeCopierOptions);
+    Metrics metrics = copier.copy();
+
+    LOG
+        .info("Avro schema '{} byes' has been copied from '{}' to '{}'", metrics.getBytesReplicated(), sourceLocation,
```
Comment (massdosage) [inline circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/util/SchemaCopier.java]:
```suggestion
        .info("Avro schema '{} bytes' has been copied from '{}' to '{}'", metrics.getBytesReplicated(), sourceLocation,
```

File: circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolverTest.java (hunk 1 above)
Comment (massdosage) [reply circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolverTest.java]: I guess previously if there was no schema it would assume that HDFS was in use whereas s3 paths always start with s3. I think this is usually the case with our on-prem setup so I'm not sure if the change you made to always check `hdfs` will work for those cases...

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/ReplicationFactoryImpl.java [hunk 9]
```diff
@@ -50,13 +47,13 @@
   public ReplicationFactoryImpl(
       SourceFactory sourceFactory,
       ReplicaFactory replicaFactory,
-      CopierFactoryManager copierFactoryManager,
+      CopierFactoryManager CopierFactoryManager,
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/ReplicationFactoryImpl.java]:
```suggestion
      CopierFactoryManager copierFactoryManager,
```

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+table-replications:
+  - source-table:
+      database-name: ${circus-train-runner.database-name}
+      table-name: ct_table_u
+    replica-table:
+      table-name: ct_table_u_copy
+      table-location: s3a://replica/${circus-train-runner.database-name}/ct_table_u_copy
+security:
+  credential-provider: jceks://file/${config-location}/aws.jceks
+
```
Comment (massdosage) [inline circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: Not sure why GH is moaning about no newline here?

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+table-replications:
+  - source-table:
+      database-name: ${circus-train-runner.database-name}
+      table-name: ct_table_u
+    replica-table:
+      table-name: ct_table_u_copy
+      table-location: s3a://replica/${circus-train-runner.database-name}/ct_table_u_copy
```
Comment (massdosage) [inline circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: Is there a reason we are using `s3a` here rather than just `s3`?

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/unpartitioned-single-table-avro-schema.yml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+table-replications:
+  - source-table:
+      database-name: ${circus-train-runner.database-name}
+      table-name: ct_table_u
+    replica-table:
+      table-name: ct_table_u_copy
+      table-location: s3a://replica/${circus-train-runner.database-name}/ct_table_u_copy
```
Comment (massdosage) [inline circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: Ditto

File: circus-train-s3-mapreduce-cp-copier/src/test/java/com/hotels/bdp/circustrain/s3mapreducecpcopier/S3MapReduceCpCopierTest.java (hunk 3 above)
Comment (massdosage) [reply circus-train-s3-mapreduce-cp-copier/src/test/java/com/hotels/bdp/circustrain/s3mapreducecpcopier/S3MapReduceCpCopierTest.java]: :)

File: circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierOptions.java [hunk 13]
```diff
@@ -21,6 +21,10 @@
 
   String IGNORE_MISSING_PARTITION_FOLDER_ERRORS = "ignore-missing-partition-folder-errors";
 
+  // internal option used to track if the destination of the replication should be treated as a folder or file. Value
+  // can be parsed with Boolean.parseValue. If not set a folder is assumed.
+  String COPY_DESTINATION_IS_FILE = "copy-destination-is-file";
```
[the remark this answers is not in the document]
Comment (massdosage) [reply circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierOptions.java]: Or is this only used internally so we wouldn't want to advertise it?

File: circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/conf/TableReplicationTest.java [hunk 14]
```diff
@@ -89,4 +93,26 @@ public void nullTableLocation() throws Exception {
     assertThat(violations.size(), is(1));
   }
 
+  @Test
+  public void mergeNullOptions() {
+    Map<String, Object> mergedCopierOptions = tableReplication.getMergedCopierOptions(null);
+    assertThat(mergedCopierOptions, is(not(nullValue())));
```
Comment (nvitucci) [inline circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/conf/TableReplicationTest.java]: A general question: is there a reason to prefer the generic `assertThat` to more specific methods, like `assertNotNull` in this case?

File: circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/conf/TableReplicationTest.java [hunk 15]
```diff
@@ -89,4 +93,26 @@ public void nullTableLocation() throws Exception {
...
+  public void mergeOptions() {
+    Map<String, Object> globalOptions = new HashMap<>();
+    globalOptions.put("one", Integer.valueOf(1));
+    globalOptions.put("two", Integer.valueOf(2));
+    Map<String, Object> overrideOptions = new HashMap<>();
+    overrideOptions.put("two", "two");
+    overrideOptions.put("three", "three");
+    tableReplication.setCopierOptions(overrideOptions);
+    Map<String, Object> mergedCopierOptions = tableReplication.getMergedCopierOptions(globalOptions);
+    assertThat((Integer) mergedCopierOptions.get("one"), is(Integer.valueOf(1)));
```
Comment (nvitucci) [inline circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/conf/TableReplicationTest.java]: Ditto: why not `assertEquals`?

**Comment by patduin** [comment]
> Looks good so far, might be an issue with that `hdfs` check on the scheme, we'd probably have to test that out internally.
> 
> What do you mean by `The only downside I can think of is custom copiers not expecting to copy a schema file.`?

Copiers are combined into a CompositeCopier which is called under the covers. I'm thinking about a copier + DataSqueeze or something. Our copiers kind of assumed an input folder + output folder. I can imagine something like dataSqueeze assumes that as well...

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml (hunk 11 above)
Comment (patduin) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: I have no clue but it doesn't work if I use s3, so I stuck with the going examples. I suspect something in s3Proxy test framework. Not ideal haven't explored it further.

File: circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/conf/TableReplicationTest.java (hunk 14 above)
Comment (patduin) [reply circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/conf/TableReplicationTest.java]: We prefer hamcrest. I'm ambivalent either way. This was copy paste since the code was moved I moved the test code as well.

File: circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolverTest.java (hunk 1 above)
Comment (patduin) [reply circus-train-avro/src/test/java/com/hotels/bdp/circustrain/avro/util/FileSystemPathResolverTest.java]: it does this test tests the two methods separately but in our code we actually always add the schema and then add the name. So we won't call it with "/" it will be hdfs:// or file:// depending on the HiveConf.
I think the code as it is now where we only add the name node if and only if we're certain that we're on hdfs makes more sense.

File: circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/transformation/AbstractAvroSerDeTransformation.java [hunk 16]
```diff
@@ -85,8 +91,15 @@ public void tableReplicationStart(EventTableReplication tableReplication, String
   }
 
   @Override
-  public void tableReplicationSuccess(EventTableReplication tableReplication, String eventId) {}
+  public void tableReplicationSuccess(EventTableReplication tableReplication, String eventId) {
+    this.eventId = null;
```
Comment (abhimanyugupta07) [inline circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/transformation/AbstractAvroSerDeTransformation.java]: Why was this changed ?

Comment (patduin) [reply circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/transformation/AbstractAvroSerDeTransformation.java]: Just resetting the values after a completed replication. Once the replication is done these value are not valid anymore another replication might come in. This state should be refreshed for every replication.

Comment (patduin) [reply circus-train-avro/src/main/java/com/hotels/bdp/circustrain/avro/transformation/AbstractAvroSerDeTransformation.java]: I stand correct I shouldn't have changed this all the tests started failing. I've put this back I guess it's ok that we override on start.

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml (hunk 10 above)
Comment (patduin) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: I think there was some trailing whitespace.

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/unpartitioned-single-table-avro-schema.yml (hunk 12 above)
Comment (patduin) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: see above I suspect s3proxy

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,5 +1,8 @@
-## [15.0.0] - 2019-11-12
+## TBD - TBD
```
Comment (massdosage) [inline CHANGELOG.md]: Should we consider this a 15.1.0 release? It's somewhere between a bug fix and a feature...

File: circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierFactoryManager.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.copier;
+
+import java.util.Map;
+
+import org.apache.hadoop.fs.Path;
+
+public interface CopierFactoryManager {
```
Comment (massdosage) [inline circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierFactoryManager.java]: CopierFactoryFactory ;)

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml (hunk 11 above)
Comment (massdosage) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainHdfsS3IntegrationTest/unpartitioned-single-table-avro-schema.yml]: OK, weird, let's leave it like this for now then.

File: circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierFactoryManager.java (hunk 18 above)
Comment (patduin) [reply circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierFactoryManager.java]: I know let's not :)

File: CHANGELOG.md (hunk 17 above)
Comment (patduin) [reply CHANGELOG.md]: ah yes forgot to bring that up. I'm for 15.1.0 as well.
